// kubra-lib-ux-components/src/atoms/stripe.js
import React from 'react';
import PropTypes from 'prop-types';
import preventDefault from '../lib/preventDefault';
import { Icon } from './icons';
import htmlProps from '../lib/htmlProps';

export function Stripe(props) {
  const { onClick, icon, content, className, id } = props;
  return (
    <div className={`stripe ${className}`} {...htmlProps(props)}>
      <a href="#stripe" onClick={preventDefault(onClick)} id={`${id}-href`}>
        <div className="stripe-container" id={`${id}-stripe-container`}>
          <div className="icon-item" id={`${id}-icon-item`}>
            <Icon name={icon} className="large" id={`${id}-${icon}`} />
          </div>
          <div className="text-item" id={`${id}-text-item`}>
            {content}
          </div>
          <div className="last-item" id={`${id}-last-item`}>
            <Icon name="chevron-right" id={`${id}-chevron-right`} />
          </div>
        </div>
      </a>
    </div>
  );
}

{
  const { node, func, string, oneOfType } = PropTypes;
  Stripe.propTypes = {
    content: oneOfType([node, string]),
    onClick: func,
    icon: string.isRequired,
    className: string,
    id: string.isRequired,
  };
}

Stripe.defaultProps = {
  content: '',
  onClick: () => {},
  className: '',
};

export default Stripe;
