// kubra-lib-ux-components/src/atoms/typography.js
import React from 'react';
import PropTypes from 'prop-types';
import { browserHistory } from 'react-router';
import { mergeClassNames } from '../lib/mergeClassNames';
import htmlProps from '../lib/htmlProps';

export function Subheader(props) {
  const { children, className } = props;
  return (
    <p {...htmlProps(props)} className={mergeClassNames('subheader', className)}>{children}</p>
  );
}

Subheader.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
};

Subheader.defaultProps = {
  className: '',
};

export function ListItem(props) {
  const { children, className } = props;
  return (
    <li {...htmlProps(props)} className={mergeClassNames('list-item', className)}>{children}</li>
  );
}

{
  const { node, string } = PropTypes;
  ListItem.propTypes = {
    children: node.isRequired,
    className: string,
  };

  ListItem.defaultProps = {
    className: '',
  };
}

export function LinkOnHover(props) {
  const { children, className, route, useHref } = props;

  const goToLink = link =>
    () => {
      if (!useHref) browserHistory.push(link);
    };

  return (
    <a
      {...htmlProps(props)}
      className={mergeClassNames('link link-on-hover', className)}
      onClick={goToLink(route)}
      href={useHref ? route : '#1'}
    >
      {children}
    </a>
  );
}

LinkOnHover.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  route: PropTypes.string,
  useHref: PropTypes.bool,
};

LinkOnHover.defaultProps = {
  className: '',
  route: '',
  useHref: true,
};

export default {
  Subheader,
  ListItem,
  LinkOnHover,
};
