// kubra-lib-ux-components/dist/atoms/modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Modal = Modal;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Basic Modal */
function Modal(_ref) {
  var isOpen = _ref.isOpen,
      children = _ref.children,
      onBackgroundClick = _ref.onBackgroundClick,
      size = _ref.size;

  return _react2.default.createElement(
    'div',
    { className: 'kubra-modal-bg ' + (isOpen ? 'open' : 'closed'), onClick: onBackgroundClick, role: 'button', tabIndex: '-1' },
    _react2.default.createElement(
      'div',
      { className: 'modal-dialog modal-' + size },
      _react2.default.createElement(
        'div',
        {
          className: 'modal-content',
          role: 'presentation',
          onClick: function onClick(e) {
            e.stopPropagation();
          }
        },
        children
      )
    )
  );
}

Modal.defaultProps = {
  // default onBackgroundClick to do nothing
  onBackgroundClick: function onBackgroundClick() {
    return null;
  },
  size: 'md'
};

Modal.propTypes = {
  /** Modal Content */
  children: _propTypes2.default.oneOfType([_propTypes2.default.element]).isRequired,
  /** Boolean to show/hide modal */
  isOpen: _propTypes2.default.bool.isRequired,
  /** Click handler for background click (.kubra-modal-bg) */
  onBackgroundClick: _propTypes2.default.func.isRequired,
  /** Preset sizes (sm, md, or lg) */
  size: _propTypes2.default.oneOf(['sm', 'md', 'lg'])
};

exports.default = {
  Modal: Modal
};
