// kubra-lib-ux-components/dist/atoms/stripe.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Stripe = Stripe;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _preventDefault = require('../lib/preventDefault');

var _preventDefault2 = _interopRequireDefault(_preventDefault);

var _icons = require('./icons');

var _htmlProps = require('../lib/htmlProps');

var _htmlProps2 = _interopRequireDefault(_htmlProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Stripe(props) {
  var onClick = props.onClick,
      icon = props.icon,
      content = props.content,
      className = props.className,
      id = props.id;

  return _react2.default.createElement(
    'div',
    Object.assign({ className: 'stripe ' + className }, (0, _htmlProps2.default)(props)),
    _react2.default.createElement(
      'a',
      { href: '#stripe', onClick: (0, _preventDefault2.default)(onClick), id: id + '-href' },
      _react2.default.createElement(
        'div',
        { className: 'stripe-container', id: id + '-stripe-container' },
        _react2.default.createElement(
          'div',
          { className: 'icon-item', id: id + '-icon-item' },
          _react2.default.createElement(_icons.Icon, { name: icon, className: 'large', id: id + '-' + icon })
        ),
        _react2.default.createElement(
          'div',
          { className: 'text-item', id: id + '-text-item' },
          content
        ),
        _react2.default.createElement(
          'div',
          { className: 'last-item', id: id + '-last-item' },
          _react2.default.createElement(_icons.Icon, { name: 'chevron-right', id: id + '-chevron-right' })
        )
      )
    )
  );
}

{
  var node = _propTypes2.default.node,
      func = _propTypes2.default.func,
      string = _propTypes2.default.string,
      oneOfType = _propTypes2.default.oneOfType;

  Stripe.propTypes = {
    content: oneOfType([node, string]),
    onClick: func,
    icon: string.isRequired,
    className: string,
    id: string.isRequired
  };
}

Stripe.defaultProps = {
  content: '',
  onClick: function onClick() {},
  className: ''
};

exports.default = Stripe;
