// kubra-lib-ux-components/src/atoms/form/MultiSelectField/DropdownSearch.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autobind from 'class-autobind';
import _ from 'lodash';
import getIdWithSuffix from '../../../lib/getIdWithSuffix';

export class DropdownSearch extends Component {
  constructor(props) {
    super(props);
    autobind(this);
  }
  componentWillUnmount() {
    this.unBindKeyListeners();
  }
  onClick(event) {
    event.stopPropagation();
    this.props.onClick(event);
  }
  onFocus(event) {
    this.props.onFocus(event);
    this.bindKeyListeners();
  }
  onBlur(event) {
    this.props.onBlur(event);
    this.unBindKeyListeners();
  }
  onKeyListen(event) {
    switch (event.keyCode) {
      case 38: // up arrow
        this.props.onUpArrow();
        return false;
      case 40: // down arrow
        this.props.onDownArrow();
        return false;
      case 9: // tab
      case 13: // enter
        event.preventDefault();
        event.stopPropagation();
        this.props.onEnter();
        return false;
      case 8: // delete
        this.props.onDelete();
        return false;
      case 27:
        this.props.onEscape();
        return false;
      default:
        break;
    }
    return false;
  }
  bindKeyListeners() {
    if (this.searchBox) {
      this.searchBox.addEventListener('keydown', this.onKeyListen);
    }
  }
  unBindKeyListeners() {
    if (this.searchBox) {
      this.searchBox.removeEventListener('keydown', this.onKeyListen);
    }
  }
  refCallback(ref) {
    this.searchBox = ref;
    this.props.refCallback(ref);
  }
  render() {
    return (
      <input
        type="text"
        ref={this.refCallback}
        placeholder={this.props.placeholder}
        className={this.props.className}
        value={this.props.value}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        onClick={this.onClick}
        onChange={this.props.onChange}
        id={getIdWithSuffix(this.props, `input-${_.kebabCase(_.camelCase(this.props.placeholder))}`)}
      />
    );
  }
}

{
  const { func, string } = PropTypes;
  DropdownSearch.propTypes = {
    refCallback: func.isRequired,
    onClick: func.isRequired,
    onChange: func.isRequired,
    onFocus: func.isRequired,
    onBlur: func.isRequired,
    placeholder: string.isRequired,
    value: string.isRequired,
    className: string,
    onEnter: func,
    onUpArrow: func,
    onDownArrow: func,
    onDelete: func,
    onEscape: func,
  };
}

DropdownSearch.defaultProps = {
  className: '',
  onEnter() {},
  onUpArrow() {},
  onDownArrow() {},
  onDelete() {},
  onEscape() {},
};

export default DropdownSearch;
