// kubra-lib-ux-components/dist/molecules/tab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabItem = TabItem;
exports.TabCloud = TabCloud;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _mergeClassNames = require('../lib/mergeClassNames');

var _htmlProps = require('../lib/htmlProps');

var _htmlProps2 = _interopRequireDefault(_htmlProps);

var _icons = require('../atoms/icons');

var _getIdWithSuffix = require('../lib/getIdWithSuffix');

var _getIdWithSuffix2 = _interopRequireDefault(_getIdWithSuffix);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function TabItem(props) {
  var onClick = props.onClick,
      label = props.label,
      disabled = props.disabled,
      className = props.className,
      icon = props.icon,
      isActive = props.isActive;

  var isActiveClass = isActive ? 'active' : '';

  return _react2.default.createElement(
    'li',
    Object.assign({
      id: (0, _getIdWithSuffix2.default)(props, 'li'),
      onClick: onClick,
      role: 'menuitem',
      className: (0, _mergeClassNames.mergeClassNames)('tab-item ' + isActiveClass, className)
    }, (0, _htmlProps2.default)(props), {
      disabled: disabled
    }),
    _react2.default.createElement(_icons.Icon, { className: icon + ' small', name: icon }),
    _react2.default.createElement(
      'span',
      { className: 'label' },
      label
    )
  );
}

TabItem.propTypes = {
  onClick: _propTypes2.default.func.isRequired,
  label: _propTypes2.default.string.isRequired,
  disabled: _propTypes2.default.bool,
  className: _propTypes2.default.string,
  isActive: _propTypes2.default.bool,
  icon: _propTypes2.default.string
};

TabItem.defaultProps = {
  disabled: false,
  onSelectValue: function onSelectValue() {},

  className: '',
  activeClassName: '',
  icon: '',
  isActive: false
};

function TabCloud(props) {
  var children = props.children,
      className = props.className,
      topPadding = props.topPadding;

  return _react2.default.createElement(
    'div',
    Object.assign({
      className: (0, _mergeClassNames.mergeClassNames)('tab-cloud', className)
    }, (0, _htmlProps2.default)(props)),
    _react2.default.createElement('div', { className: 'top-padding', style: { flex: '0 0 ' + topPadding } }),
    _react2.default.createElement(
      'div',
      { className: 'tab-container' },
      children
    )
  );
}

TabCloud.propTypes = {
  children: _propTypes2.default.node.isRequired,
  className: _propTypes2.default.string,
  topPadding: _propTypes2.default.string
};

TabCloud.defaultProps = {
  className: '',
  topPadding: null
};

exports.default = {
  TabItem: TabItem,
  TabCloud: TabCloud
};
