// kubra-lib-ux-components/src/atoms/button/ToggleButton.js
import React, { Component } from 'react';
import _ from 'lodash';
import autobind from 'class-autobind';
import PropTypes from 'prop-types';
import htmlProps from '../../lib/htmlProps';

function oppositeLabelFloat(value) {
  return value.toLowerCase() === 'left' ? 'right' : 'left';
}

export class ToggleButton extends Component {
  constructor(props) {
    super(props);
    this.state = this.defaultState();
    autobind(this);
  }

  componentWillReceiveProps({ value }) {
    if (value === undefined) {
      return;
    }
    this.setState(() => ({ value }));
  }

  onSelect(value) {
    this.setState(() => ({ value }));
  }

  defaultState() {
    return {
      value: this.props.value || false,
    };
  }

  render() {
    const {
      name,
      onClick,
      label,
      labelFloat,
    } = this.props;

    const {
      onSelect,
    } = this;

    const _onChange = (value) => {
      if (value !== this.state.value) {
        onSelect(value);
        onClick(value);
      }
    };

    return (
      <div>
        { !_.isNil(label) &&
        <span className="toggle-label" style={{ float: `${labelFloat}` }}>{label}</span>
        }
        <span className="toggle-cont" style={{ float: oppositeLabelFloat(labelFloat) }}>
          <input
            {...htmlProps(this.props)}
            name={name}
            id={`toggle-input-${_.kebabCase(name)}`}
            className="toggle flat"
            type="checkbox"
            onChange={() => _onChange(!this.state.value)}
            checked={this.state.value}
          />
          <label
            className="toggle-button"
            htmlFor={`toggle-input-${_.kebabCase(name)}`}
          />
        </span>
      </div>
    );
  }
}

const { string, func, bool } = PropTypes;
ToggleButton.propTypes = {
  name: string,
  value: bool,
  label: string,
  labelFloat: string,
  onClick: func,
};

ToggleButton.defaultProps = {
  name: 'test',
  value: false,
  label: null,
  labelFloat: 'left',
  onClick() {},
};

export default {
  ToggleButton,
};
