// kubra-lib-ux-components/src/molecules/confirmDelete.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autobind from 'class-autobind';
import _ from 'lodash';
import { ButtonSecondary, ButtonPrimary } from '../atoms/buttons';
import { TextField } from '../atoms/form/TextField';
import { Modal } from '../atoms/modal';

/** Basic ConfirmDelete */
export class ConfirmDelete extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userInput: '',
      class: '',
    };
    autobind(this);
  }
  onChange(e) {
    this.setState({ userInput: e });
  }
  handleConfirm() {
    this.props.handleDelete();
    this.setState({ class: '' });
  }
  handleError() {
    this.setState({ class: 'invalid' });
  }
  render() {
    const {
      isOpen,
      dataType,
      clickedElem,
      confirmType,
      onBackgroundClick } = this.props;
    return (
      <div>
        <Modal
          isOpen={isOpen}
          size="md"
          onBackgroundClick={onBackgroundClick}
        >
          <div style={{ padding: '10px 20px' }}>
            <h1>Confirm Delete</h1>
            <p>Are you sure you want to delete this {_.startCase(_.toLower(dataType))}?</p>
            {confirmType === 'input' &&
              <TextField
                name="confirmDelete"
                hintText={`Enter "${clickedElem}" to Delete`}
                onChange={this.onChange}
                className={this.state.class}
                label={`${_.startCase(_.toLower(dataType))} Name`}
              />
            }
            <div style={{ textAlign: 'right', padding: '10px 20px' }}>
              <ButtonSecondary
                label="Cancel"
                onClick={onBackgroundClick}
              />
              <span style={{ paddingRight: '16px' }} />
              <ButtonPrimary
                label="Confirm"
                onClick={
                  confirmType !== 'input' ||
                  clickedElem === this.state.userInput ?
                  this.handleConfirm : this.handleError}
              />
            </div>
          </div>
        </Modal>
      </div>
    );
  }
}

ConfirmDelete.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onBackgroundClick: PropTypes.func.isRequired,
  dataType: PropTypes.string.isRequired,
  clickedElem: PropTypes.string.isRequired,
  confirmType: PropTypes.string,
  handleDelete: PropTypes.func.isRequired,
};

ConfirmDelete.defaultProps = {
  confirmType: '',
};

export default {
  ConfirmDelete,
};
