// kubra-lib-ux-components/src/molecules/pagination.js
import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { mergeClassNames } from '../lib/mergeClassNames';
import { IconButton } from '../atoms/buttons';
import htmlProps from '../lib/htmlProps';

export function Pagination(props) {
  const { onClick, disabled, className, currentPage, totalPages, alwaysShow, id } = props;
  const classNames = mergeClassNames('pagination', className);

  const pageClicked = index => () => {
    if (index < 0 || index >= totalPages || disabled) return;
    onClick(index);
  };

  if (!alwaysShow && totalPages < 2) {
    return (<div />);
  }

  return (
    <div className={classNames} {...htmlProps}>
      <IconButton
        icon={'chevron-left'}
        onClick={pageClicked(currentPage - 1)}
        className="small"
        id={`${id}pagination-previous`}
        disabled={currentPage - 1 < 0}
      />
      {_.times(totalPages, ((index) => {
        const pageIndex = index + 1;
        return (
          <a
            href={`#${pageIndex}`}
            onClick={pageClicked(index)}
            className={`pagination-link ${index === currentPage ? 'active' : ''}`}
            id={`${id}page-${pageIndex}`}
            key={pageIndex}
          >
            {pageIndex}
          </a>
        );
      }))}
      <IconButton
        icon={'chevron-right'}
        onClick={pageClicked(currentPage + 1)}
        className="small"
        id={`${id}pagination-next`}
        disabled={currentPage + 1 >= totalPages}
      />
    </div>
  );
}

Pagination.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  id: PropTypes.string,
  currentPage: PropTypes.number.isRequired,
  totalPages: PropTypes.number.isRequired,
  disabled: PropTypes.bool,
  alwaysShow: PropTypes.bool,
};

Pagination.defaultProps = {
  className: '',
  id: '',
  disabled: false,
  alwaysShow: false,
};

export default {
  Pagination,
};
